json_object.c

```diff
@@ -901,18 +903,23 @@ static int json_object_double_to_json_string_format(struct json_object *jso, str
 	 * ECMA 262 section 9.8.1 defines
 	 * how to handle these cases as strings
 	 */
+#ifdef HAVE_DECL_ISNAN
 	if (isnan(jso->o.c_double))
 	{
 		size = snprintf(buf, sizeof(buf), "NaN");
 	}
-	else if (isinf(jso->o.c_double))
+	else
+#endif
+#ifdef HAVE_DECL_ISINF
+	if (isinf(jso->o.c_double))
```

Comment:
eh?  These systems have `INFINITY` (as referenced in json_tokener.c), but no way to check if a number is that value?

Reply:
Correct. On IBM i the printf function will automatically turn a value of INFINITY to "infinity", and NAN to "nan". There are no isinf or isnan functions.

Reply:
for what it's worth, the python json decoder will accept `{"val": Infinity}` and `{"val": NaN}`, but will not accept `{"val": infinity}` or `{"val": nan}` (and ECMA 262 dictates the capitalized versions)

Reply:
hmm... that's not going to work so well with the tests, since they're going to be looking for "Infinity" and "NaN".
Are you sure they aren't available if you include float.h?
Or, perhaps _XOPEN_SOURCE needs to be defined? (e.g. like mentioned at https://github.com/matplotlib/matplotlib/issues/12535/)

Reply:
If they're not provided, we'll need to provide our own.  Can you please test this program and see if it works?:
```
#include <stdlib.h>
#include <stdio.h>
#include <float.h>

int main()
{
    double f[] = {
        0.0d, 1.0d, DBL_MAX, 0.0d / 0.0d, 1.0d/0.0d, -1.0d/0.0d
    };
    int i;

    for (i = 0; i < 6; i++) {
        double x = f[i];

        int is_nan = (x != x);
        int is_inf = (x < -DBL_MAX || x > DBL_MAX);

        printf("%20g%4d%4d\n", x, is_nan, is_inf);
    }

    return 0;
}
```

It should produce output like:
```
                   0   0   0
                   1   0   0
        1.79769e+308   0   0
                -nan   1   0
                 inf   0   1
                -inf   0   1
```

Reply:
There's definitely no isnan or isinf on IBM i according to the Knowledge Center (https://www.ibm.com/support/knowledgecenter/ssw_ibm_i_74/rtref/rtrefmain.htm). I will run your test program on IBM i and let you know the results when they are available.

Reply:
Your original program crashed. I modified it to look like this:
const unsigned int lnIND[2] = {0x00000000, 0xfff80000};
const unsigned int lnINF[2] = {0x00000000, 0x7ff00000};

int main()
{
  const double ZERO                 = 0.0;
  const double ONE                  = 1.0;
  const double AN_INDETERMINATE     = *( double* )lnIND;
  const double AN_INFINITY_POSITIVE = *( double* )lnINF;
  const double AN_INFINITY_NEGATIVE = -AN_INFINITY_POSITIVE;

  double f[] = {ZERO, ONE, DBL_MAX, AN_INDETERMINATE, AN_INFINITY_POSITIVE, AN_INFINITY_NEGATIVE};

It produced this output:
```
                   0   0   0
                   1   0   0
        1.79769e+308   0   0
        2.12174e-314   0   0
        1.06048e-314   0   0
       -1.06048e-314   0   0
```

Reply:
I'm not sure how else to get it to produce a Nan or a value for Infinity.

Reply:
Crashes how?  Do the machines you're on not properly handle floating point divide-by-zero traps?  Is it defined to use something other than the IEEE 754 standard for floating point math?

Reply:
Also, it seems like you haven't managed to actually get real NaN and Infinity values in the test program, since you said that printf would emit "nan" or "infinity", but that isn't happening.
How about trying to use the `INFINITY` and `NAN` constants?  You might need to include math.h to get those.

Reply:
Using INFINITY and NAN I get this:
```
                    0   0   0  
                    1   0   0  
         1.79769e+308   0   0  
                  nan   1   0  
             infinity   0   1  
            -infinity   0   1  
```

Reply:
Perfect!  I'll merge this PR as-is, then fix things up afterwards.